### PR TITLE
Add global flag (`-C`) to change execution directory

### LINKED ISSUE #448: $ gem install should look for dependencies in the same directory as the dependant gem
If I want to install a.gem and it depends on b.gem and I want to install them from local files(`/tmp/gemstage/a.gem` and `/tmp/gemstage/b.gem`), I should be able to install them as follows:

```
gem install /tmp/gemstage/a.gem --local
```

But this will fail because it won't know where to look for b.gem.  I can cd to `/tmp/gemstage/` and then run:

```
gem install a.gem --local
```

It would be a nice feature if gem install would look for dependencies in the same folder as the gem you're installing. That way you won't have to be concerned with changing directories when scripting these installations.

### PR DESCRIPTION
**Pull request by gustavothecoder**
## What was the end-user or developer problem that led to this PR?

`gem install`, when installing a `.gemspec` file directly, looks for dependencies of the gemspec in other gemspec files present in the CWD. This works when the target `.gemspec` is in the CWD but not if it's somewhere else.

So if I have `a.gemspec` and `b.gemspec`, where `b` is a dependency if `a`, at folder `foo/bar`, then `cd foo/bar; gem install a.gemspec` works but `gem install foo/bar/a.gemspec` does not work.

## What is your fix for the problem, implemented in this PR?

This PR does not fix the issue per se, but introduces a `-C` flag to simplify making this work, by running, `gem -C foo/bar a.gemspec`. So we'll make this PR tentatively close #448 and wait for feedback in case someone still needs #448 itself after this.

As discussed before (#448 and #4058), and explained above, this solution introduces a flag (`-C`) that gives users the ability to change the execution directory from the current one to another. This flag was inspired by [git -C](https://git-scm.com/docs/git#Documentation/git.txt--Cltpathgt) and [ruby -C](https://linux.die.net/man/1/ruby).

Although the problem reported in #448 is related to `gem install`, implementing `-C` as a global flag provides consistency with other command line tools and exposes this feature to other commands.

`gem build` already has this same option, so I added a deprecation warning for it.

Closes #448.

## Make sure the following tasks are checked

- [x] Describe the problem / feature
- [x] Write [tests](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#tests) for features and bug fixes
- [x] Write code to solve the problem
- [x] Make sure you follow the [current code style](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#code-formatting) and [write meaningful commit messages without tags](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#commit-messages)

### CODE REVIEW

File: lib/rubygems/command.rb [hunk 1]
```diff
@@ -630,7 +630,7 @@ def wrap(text, width) # :doc:
   Usage:
     gem -h/--help
     gem -v/--version
-    gem command [arguments...] [options...]
+    gem [-C PATH] command [arguments...] [options...]
```
Comment (gustavothecoder) [inline lib/rubygems/command.rb:633]: Is there a better way to document this type of option?

File: lib/rubygems/commands/build_command.rb [hunk 2]
```diff
@@ -26,6 +26,8 @@ def initialize
     add_option "-C PATH", "Run as if gem build was started in <PATH> instead of the current working directory." do |value, options|
       options[:build_path] = value
     end
+    deprecate_option "-C",
+                     extra_msg: "-C is a global flag now. Use `gem -C PATH build GEMSPEC_FILE [options]` instead"
```
Comment (gustavothecoder) [inline lib/rubygems/commands/build_command.rb:30]: If you guys want, I can add a version here.

File: test/rubygems/test_gem_command_manager.rb [hunk 3]
```diff
@@ -126,6 +126,46 @@ def test_run_crash_command
...
+    custom_start_point.tr!("_", "-")
+
+    use_ui @ui do
+      assert_raise Gem::MockGemUi::TermError do
+        @command_manager.process_args %W[-C #{custom_start_point} install net-scp-4.0.0.gem --local]
+      end
+    end
+
+    assert_match(/#{custom_start_point} isn't a directory./i, @ui.error)
+  end
```
Comment (gustavothecoder) [inline test/rubygems/test_gem_command_manager.rb:167]: My test approach only cares about the `-C` behavior. Is this sufficient or we need to write test to all comands?

File: lib/rubygems/command.rb (hunk 1 above)
Comment (deivid-rodriguez) [reply lib/rubygems/command.rb:633]: I guess we could add a "Global options:" section below "Usage:"? What do you think?

File: lib/rubygems/commands/build_command.rb (hunk 2 above)
Comment (deivid-rodriguez) [reply lib/rubygems/commands/build_command.rb:30]: Let's say 4.0 I guess.

File: test/rubygems/test_gem_command_manager.rb (hunk 3 above)
Comment (deivid-rodriguez) [reply test/rubygems/test_gem_command_manager.rb:167]: What you did seems fine to me!

File: lib/rubygems/command.rb (hunk 1 above)
Comment (gustavothecoder) [reply lib/rubygems/command.rb:633]: Good idea.

**Comment by deivid-rodriguez** [comment]
Thanks @gustavothecoder, it looks very good!

I'm releasing RubyGems 3.4.0 now, but will merge & release this one with a future release (RubyGems 3.4.1 hopefully!).

**Review by deivid-rodriguez — COMMENTED** [review]
Looks good, only thing I'm missing is maybe a test for the deprecation of `gem build -C`.

**Comment by gustavothecoder** [comment]
> Looks good, only thing I'm missing is maybe a test for the deprecation of `gem build -C`.

Added!

**Review by deivid-rodriguez — APPROVED** [review]
😍

**Comment by deivid-rodriguez** [comment]
@simi Any opinion on this? In particular, you've been against shipping deprecations with patch level releases. I _believe_ for this relatively minor usage it's fine, but if not, we can split this PR, add the enhancement to the next version, and leave the deprecation for next year's release.

**Comment by simi** [comment]
> @simi Any opinion on this? In particular, you've been against shipping deprecations with patch level releases. I _believe_ for this relatively minor usage it's fine, but if not, we can split this PR, add the enhancement to the next version, and leave the deprecation for next year's release.

This seems ok to me.

**Review by simi — APPROVED** [review]

**Comment by deivid-rodriguez** [comment]
Awesome, thanks @simi and @gustavothecoder! ❤️

**Comment by deivid-rodriguez** [comment]
This PR will close the oldest issue we currently have opened 💪.

**Pull request merged**